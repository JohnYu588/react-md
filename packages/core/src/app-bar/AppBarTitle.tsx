import { forwardRef } from "react";

import { type TextOverflow } from "../cssUtils.js";
import { Typography, type TypographyProps } from "../typography/Typography.js";
import { type TypographyType } from "../typography/typographyStyles.js";
import { type AppBarTitleClassNameOptions, appBarTitle } from "./styles.js";

/**
 * @since 6.0.0 The `keyline` prop was changed from a boolean to a type
 * union of different keylines: {@link AppBarTitleKeyline}
 * @since 6.0.0 The `noWrap` prop was removed in favor of the `textOverflow`
 * prop inherited through the base `Typography` component. The `textOverflow`
 * will default to `"ellipsis"` which is new as well.
 */
export interface AppBarTitleProps
  extends TypographyProps,
    AppBarTitleClassNameOptions {
  /** @defaultValue `"headline-6"` */
  type?: TypographyType;

  /**
   * @defaultValue `"ellipsis"`
   */
  textOverflow?: TextOverflow;
}

/**
 * @example Updating the Keyline
 * ```tsx
 * import { AppBar } from "@react-md/core/app-bar/AppBar";
 * import { AppBarTitle } from "@react-md/core/app-bar/AppBarTitle";
 * import type { ReactElement } from "react";
 *
 * function Example(): ReactElement {
 *   return (
 *     <AppBar>
 *       <AppBarTitle keyline="nav">
 *         Offset as if there was a nav button to the left
 *       </AppBarTitle>
 *     </AppBar>
 *   );
 * }
 * ```
 *
 * @since 6.0.0 The `keyline` prop was changed from a boolean to a type
 * union of different keylines: {@link AppBarTitleKeyline}
 * @since 6.0.0 The `noWrap` prop was removed in favor of the `textOverflow`
 * prop inherited through the base `Typography` component. The `textOverflow`
 * will default to `"ellipsis"` which is new as well.
 */
export const AppBarTitle = forwardRef<HTMLHeadingElement, AppBarTitleProps>(
  function AppBarTitle(props, ref) {
    const {
      type = "headline-6",
      keyline = "small",
      children,
      className,
      textOverflow = "ellipsis",
      ...remaining
    } = props;
    return (
      <Typography
        {...remaining}
        ref={ref}
        type={type}
        className={appBarTitle({
          keyline,
          className,
        })}
        textOverflow={textOverflow}
      >
        {children}
      </Typography>
    );
  }
);
