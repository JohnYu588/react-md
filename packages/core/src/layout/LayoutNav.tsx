"use client";

import { cnb } from "cnbuilder";
import { type HTMLAttributes, type ReactNode, forwardRef } from "react";

import {
  type BaseSheetClassNameOptions,
  DEFAULT_SHEET_CLASSNAMES,
  DEFAULT_SHEET_TIMEOUT,
  sheet,
} from "../sheet/styles.js";
import {
  type CSSTransitionClassNames,
  type TransitionActions,
  type TransitionCallbacks,
  type TransitionTimeout,
} from "../transition/types.js";
import { useCSSTransition } from "../transition/useCSSTransition.js";
import { layoutNav } from "./layoutNavStyles.js";

/**
 * @since 6.0.0
 */
export interface LayoutNavProps
  extends HTMLAttributes<HTMLDivElement>,
    BaseSheetClassNameOptions,
    TransitionCallbacks,
    TransitionActions {
  children: ReactNode;

  /**
   * The component to render as.
   *
   * @defaultValue `"nav"`
   */
  as?: "nav" | "div";

  /**
   * Set this to `true` to display the navigation.
   */
  expanded: boolean;

  /**
   * Set this to `true` to force the navigation to appear below the fixed app
   * bar.
   *
   * @defaultValue `false`
   */
  appBarOffset?: boolean;

  /** @defaultValue {@link DEFAULT_SHEET_TIMEOUT} */
  timeout?: TransitionTimeout;
  /** @defaultValue {@link DEFAULT_SHEET_CLASSNAMES} */
  classNames?: CSSTransitionClassNames;
}

/**
 * **Client Component**
 *
 * This component is used to render a persistent navigation on the left of the
 * page. There is some built-in functionality to support resizing with the
 * `LayoutWindowSplitter` and expandable layouts. Do not use this component for
 * temporary navigation and instead use a `Sheet`.
 *
 * @example Full Height Layout
 * ```tsx
 * "use client";
 * import { LayoutAppBar } from "@react-md/core/layout/LayoutAppBar";
 * import { LayoutNav } from "@react-md/core/layout/LayoutNav";
 * import { Main } from "@react-md/core/layout/Main";
 * import { useHorizontalLayoutTransition } from "@react-md/core/layout/useHorizontalLayoutTransition";
 * import type { ReactElement, PropsWithChildren } from "react";
 *
 * function Layout({ children }: PropsWithChildren): ReactElement {
 *   const { elementProps } = useHorizontalLayoutTransition({
 *     transitionIn: true,
 *   });
 *
 *   return (
 *     <>
 *       <LayoutAppBar {...elementProps}>
 *         <YourAppBarContent />
 *       </LayoutAppBar>
 *       <LayoutNav expanded>
 *         <YourNavigationComponent />
 *       </LayoutNav>
 *       <Main navOffset appBarOffset {...elementProps}>
 *         {children}
 *       </Main>
 *     </>
 *   );
 * }
 * ```
 *
 * @example Toggleable Layout
 * ```tsx
 * "use client";
 * import { LayoutAppBar } from "@react-md/core/layout/LayoutAppBar";
 * import { Main } from "@react-md/core/layout/Main";
 * import { useHorizontalLayoutTransition } from "@react-md/core/layout/useHorizontalLayoutTransition";
 * import { useToggle } from "@react-md/core/useToggle";
 * import MenuIcon from "@react-md/material-icons/MenuIcon";
 * import { cnb } from "cnbuilder";
 * import type { ReactElement, PropsWithChildren } from "react";
 *
 * function Layout({ children }: PropsWithChildren): ReactElement {
 *   const { toggled: expanded, toggle } = useToggle();
 *   const { elementProps } = useHorizontalLayoutTransition({
 *     transitionIn: expanded,
 *   });
 *
 *   return (
 *     <>
 *       <LayoutAppBar {...elementProps}>
 *         <Button
 *           aria-label="Navigation"
 *           onClick={toggle}
 *           buttonType="icon"
 *         >
 *           <MenuIcon />
 *         </Button>
 *         <YourAppBarContent />
 *       </LayoutAppBar>
 *       <LayoutNav expanded>
 *         <YourNavigationComponent />
 *       </LayoutNav>
 *       <Main navOffset={expanded} appBarOffset {...elementProps}>
 *         {children}
 *       </Main>
 *     </>
 *   );
 * }
 * ```
 *
 * @since 6.0.0
 */
export const LayoutNav = forwardRef<HTMLDivElement, LayoutNavProps>(
  function LayoutNav(props, ref) {
    const {
      as: Component = "nav",
      "aria-labelledby": ariaLabelledBy,
      "aria-label": ariaLabel = Component == "nav" && !ariaLabelledBy
        ? "Navigation"
        : undefined,
      expanded,
      children,
      className,
      timeout = DEFAULT_SHEET_TIMEOUT,
      classNames = DEFAULT_SHEET_CLASSNAMES,
      appear,
      enter,
      exit,
      onEnter,
      onEntering,
      onEntered,
      onExit,
      onExited,
      onExiting,
      appBarOffset,
      ...remaining
    } = props;
    const { elementProps } = useCSSTransition({
      nodeRef: ref,
      timeout,
      className: cnb(
        layoutNav({ appBarOffset }),
        sheet({
          className,
          horizontalSize: "none",
          disableOverlay: true,
        })
      ),
      classNames,
      enter,
      exit,
      appear,
      onEnter,
      onEntering,
      onEntered,
      onExit,
      onExited,
      onExiting,
      exitedHidden: true,
      transitionIn: expanded,
    });

    return (
      <Component
        aria-label={ariaLabel}
        aria-labelledby={ariaLabelledBy}
        {...remaining}
        {...elementProps}
      >
        {children}
      </Component>
    );
  }
);
