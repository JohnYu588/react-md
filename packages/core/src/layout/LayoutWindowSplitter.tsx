"use client";

import { forwardRef, useMemo } from "react";

import { type DefinedCSSVariableName } from "../theme/types.js";
import { useCSSVariables } from "../theme/useCSSVariables.js";
import {
  type BaseWindowSplitterProps,
  WindowSplitter,
} from "../window-splitter/WindowSplitter.js";
import { type WindowSplitterWidgetProps } from "../window-splitter/useWindowSplitter.js";
import {
  type LayoutWindowSplitterClassNameOptions,
  layoutWindowSplitter,
} from "./layoutWindowSplitterStyles.js";

/**
 * @since 6.0.0
 */
export interface LayoutWindowSplitterProps
  extends WindowSplitterWidgetProps<HTMLButtonElement>,
    Omit<
      BaseWindowSplitterProps,
      keyof WindowSplitterWidgetProps<HTMLButtonElement>
    >,
    LayoutWindowSplitterClassNameOptions {
  /** @defaultValue `"Resize Navigation"` */
  "aria-label"?: string;

  /**
   * The current navigation width (in px).
   */
  navWidth: number;
}

/**
 * **Client Component**
 *
 * This component is used to resize the `LayoutNav` component.
 *
 * @see {@link useResizableExpandableLayout} for a default implementation for
 * this component.
 *
 * @example Dynamic Resizing
 * ```tsx
 * import { LayoutNav } from "@react-md/core/layout/LayoutNav";
 * import { LayoutWindowSplitter } from "@react-md/core/layout/LayoutWindowSplitter";
 * import { Main } from "@react-md/core/layout/Main";
 * import { NoSsr } from "@react-md/core/NoSsr";
 * import { useWindowSplitter } from "@react-md/core/window-splitter/useWindowSplitter";
 * import { useWindowSize } from "@react-md/core/useWindowSize";
 * import type { ReactElement, PropsWithChildren } from "react"
 *
 * function MyWindowSplitter(): ReactElement {
 *   const { width } = useWindowSize({ disableHeight: true });
 *   const min = 96;
 *   const max = Math.max(600, width * .7);
 *
 *   const { value, splitterProps } = useWindowSplitter({
 *     min,
 *     max,
 *     defaultValue: 256,
 *   });
 *
 *   return (
 *     <LayoutWindowSplitter
 *       aria-controls="layout-nav-id"
 *       {...splitterProps}
 *       value={value}
 *     />
 *   );
 * }
 *
 * function Layout({ children }: PropsWithChildren): ReactElement {
 *   return (
 *     <>
 *       <LayoutNav id="layout-nav" {...otherProps}>
 *         <YourNavigation />
 *       </LayoutNav>
 *       // only required when using SSR
 *       <NoSsr>
 *         <MyWindowSplitter />
 *       </NoSsr>
 *       <Main navOffset>
 *         {children}
 *       </Main>
 *     </>
 *   );
 * }
 * ```
 *
 * @since 6.0.0
 */
export const LayoutWindowSplitter = forwardRef<
  HTMLButtonElement,
  LayoutWindowSplitterProps
>(function LayoutWindowSplitter(props, ref) {
  const {
    "aria-labelledby": ariaLabelledBy,
    "aria-label": ariaLabel = ariaLabelledBy ? undefined : "Resize Navigation",
    appBarOffset,
    disableResponsive,
    className,
    navWidth,
    ...remaining
  } = props;

  useCSSVariables<DefinedCSSVariableName>(
    useMemo(
      () => [{ name: "--rmd-layout-size", value: `${navWidth}px` }],
      [navWidth]
    )
  );

  return (
    <WindowSplitter
      {...remaining}
      ref={ref}
      aria-label={ariaLabel as string}
      aria-labelledby={ariaLabelledBy}
      className={layoutWindowSplitter({
        appBarOffset,
        disableResponsive,
        className,
      })}
    />
  );
});
