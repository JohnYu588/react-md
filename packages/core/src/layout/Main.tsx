"use client";
import type { ElementType, HTMLAttributes, ReactNode, Ref } from "react";
import { forwardRef } from "react";
import { useEnsuredId } from "../useEnsuredId.js";
import type { MainClassNameOptions } from "./mainStyles.js";
import { main } from "./mainStyles.js";
import { useMainTabIndex } from "./useMainTabIndex.js";

/**
 * @remarks \@since 6.0.0
 */
export type CustomMainElement = ElementType<
  HTMLAttributes<HTMLElement> & {
    ref: Ref<HTMLElement>;
    className?: string;
    tabIndex?: number;
  }
>;

/**
 * @remarks \@since 6.0.0
 */
export interface MainProps
  extends HTMLAttributes<HTMLElement>,
    MainClassNameOptions {
  /**
   * @defaultValue `"main"`
   */
  as?: CustomMainElement;
  children: ReactNode;
}

/**
 * **Client Component**
 *
 * This component is really only used to dynamically set the `tabIndex` to `-1`
 * while using a keyboard for the `SkipToMainContent` component's focus behavior
 * to work correctly. If you don't need that functionality, use the {@link main}
 * style utility instead.
 *
 * @example
 * Styles Only
 * ```tsx
 * import { main as mainStyles } from "@react-md/core":
 *
 * function MyCustomMainElement({ children }) {
 *   return (
 *     <main
 *       className={mainStyles({
 *         navOffset: true,
 *         appBarOffset: true,
 *       })}
 *     >
 *       {children}
 *     </main>
 *   );
 * }
 * ```
 *
 * @remarks \@since 6.0.0 Renamed from `LayoutMain` removed a lot of
 * functionality to keep this component simple.
 */
export const Main = forwardRef<HTMLElement, MainProps>(
  function Main(props, ref) {
    const {
      as: Component = "main",
      id: propId,
      className,
      children,
      tabIndex: propTabIndex,
      navOffset,
      appBarOffset,
      ...remaining
    } = props;
    const id = useEnsuredId(propId, "main");
    const tabIndex = useMainTabIndex(propTabIndex);

    return (
      <Component
        {...remaining}
        id={id}
        ref={ref}
        className={main({ navOffset, appBarOffset, className })}
        tabIndex={tabIndex}
      >
        {children}
      </Component>
    );
  }
);
