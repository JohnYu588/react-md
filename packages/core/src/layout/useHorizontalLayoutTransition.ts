"use client";
import { cnb } from "cnbuilder";
import { DEFAULT_SHEET_TIMEOUT } from "../sheet/Sheet.js";
import {
  type CSSTransitionClassNames,
  type CSSTransitionHookReturnValue,
  type PreconfiguredCSSTransitionOptions,
  type TransitionTimeout,
} from "../transition/types.js";
import { useCSSTransition } from "../transition/useCSSTransition.js";

/**
 * @since 6.0.0
 */
export const DEFAULT_HORIZONTAL_LAYOUT_TRANSITION_CLASSNAMES = {
  appearDone: "rmd-layout-h--active",
  enter: "rmd-layout-h--enter",
  enterActive: "rmd-layout-h--active",
  enterDone: "rmd-layout-h--active",
  exit: "rmd-layout-h--exit",
} satisfies Readonly<CSSTransitionClassNames>;

/**
 * @since 6.0.0
 */
export interface HorizontalLayoutTransitionOptions
  extends PreconfiguredCSSTransitionOptions<HTMLElement> {
  /**
   * @see {@link DEFAULT_SHEET_TIMEOUT}
   * @defaultValue `DEFAULT_SHEET_TIMEOUT`
   */
  timeout?: TransitionTimeout;
  /**
   * @see {@link DEFAULT_HORIZONTAL_LAYOUT_TRANSITION_CLASSNAMES}
   * @defaultValue `DEFAULT_HORIZONTAL_LAYOUT_TRANSITION_CLASSNAMES`
   */
  classNames?: CSSTransitionClassNames;
}

/**
 * @example
 * ```tsx
 * import { LayoutAppBar, Main, useHorizontalLayoutTransition } from "@react-md/core";
 * import type { ReactElement, ReactNode } from "react";
 *
 * interface Props {
 *   title: ReactNode;
 *   children: ReactNode;
 * }
 *
 * export default function Example({ title, children }: Props): ReactElement {
 *   const { toggled: staticNavExpanded, toggle: toggleStaticNav } = useToggle();
 *   const { elementProps, className } = useHorizontalLayoutTransition({
 *     transitionIn: staticNavExpanded,
 *   });
 *
 *   return (
 *     <>
 *       <LayoutAppBar className={className}>
 *         <Button
 *           aria-label="Navigation"
 *           buttonType="icon"
 *           onClick={toggleStaticNav}
 *         >
 *           <MenuOutlinedIcon />
 *         '</Button>
 *         {title}
 *       </LayoutAppBar>
 *       <Main {...elementProps}>
 *         {children}
 *       </Main>
 *     </>
 *   );
 * }
 * ```
 *
 * @since 6.0.0
 */
export function useHorizontalLayoutTransition<
  E extends HTMLElement = HTMLElement,
>(options: HorizontalLayoutTransitionOptions): CSSTransitionHookReturnValue<E> {
  return useCSSTransition({
    timeout: DEFAULT_SHEET_TIMEOUT,
    classNames: DEFAULT_HORIZONTAL_LAYOUT_TRANSITION_CLASSNAMES,
    ...options,
    className: cnb("rmd-layout-h", options.className),
  });
}
