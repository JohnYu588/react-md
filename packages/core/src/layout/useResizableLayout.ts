import { useEnsuredId } from "../useEnsuredId.js";
import { DISPLAY_NONE_CLASS } from "../utils/isElementVisible.js";
import { type LayoutWindowSplitterProps } from "./LayoutWindowSplitter.js";
import {
  type ExpandableLayoutImplementation,
  type ExpandableLayoutOptions,
  type ProvidedLayoutNavProps,
} from "./useExpandableLayout.js";
import { useExpandableLayout } from "./useExpandableLayout.js";
import {
  type LayoutWindowSplitterOptions,
  useLayoutWindowSplitter,
} from "./useLayoutWindowSplitter.js";

/**
 * @since 6.0.0
 */
export interface ResizableLayoutOptions
  extends ExpandableLayoutOptions,
    LayoutWindowSplitterOptions {
  /**
   * This id will be used as the `aria-controls` prop for the
   * `LayoutWindowSplitter` and should be applied to the `LayoutNav` as an `id`.
   *
   * @defaultValue `"layout-nav" + useId()`
   */
  navId?: string;
}

/**
 * @since 6.0.0
 */
export interface ProvidedResizableLayoutNavProps
  extends ProvidedLayoutNavProps {
  id: string;
}

/**
 * @since 6.0.0
 */
export interface ResizableLayoutImplementation
  extends ExpandableLayoutImplementation {
  expandableNavProps: ProvidedResizableLayoutNavProps;
  windowSplitterProps: LayoutWindowSplitterProps;
}

/**
 * @example Main Usage
 * ```tsx
 * import { AppBarTitle } from "@react-md/core/app-bar/AppBarTitle";
 * import { Button } from "@react-md/core/button/Button";
 * import { LayoutAppBar } from "@react-md/core/layout/LayoutAppBar";
 * import { LayoutNav } from "@react-md/core/layout/LayoutNav";
 * import { LayoutWindowSplitter } from "@react-md/core/layout/LayoutWindowSplitter";
 * import { Main } from "@react-md/core/layout/Main";
 * import { useResizableLayout } from "@react-md/core/layout/useResizableLayout";
 * import { Sheet } from "@react-md/core/sheet/Sheet";
 * import { type ReactElement, type ReactNode } from "react";
 *
 * import { CustomNavigation } from "./CustomNavigation";
 *
 * export interface LayoutProps {
 *   children: ReactNode;
 * }
 *
 * export function Layout(props: LayoutProps): ReactElement {
 *   const { children } = props;
 *
 *   // choose whichever one for your app
 *   // nextjs app dir
 *   const pathname = usePathname();
 *   // nextjs pages
 *   const { pathname } = useRouter();
 *   // react router
 *   const { pathname } = useHistory();
 *
 *   const {
 *     temporary,
 *     appBarProps,
 *     expandableNavProps,
 *     mainProps,
 *     navToggleProps,
 *     temporaryNavProps,
 *     windowSplitterProps,
 *   } = useResizableLayout({ pathname });
 *
 *   return (
 *     <>
 *       <LayoutAppBar {...appBarProps}>
 *         <Button {...navToggleProps} />
 *         <AppBarTitle>Hello, world!</AppBarTitle>
 *       </LayoutAppBar>
 *       <LayoutNav {...expandableNavProps}>
 *         <CustomNavigation />
 *       </LayoutNav>
 *       <LayoutWindowSplitter {...windowSplitterProps} />
 *       {temporary && (
 *         <Sheet {...temporaryNavProps}>
 *           <CustomNavigation />
 *         </Sheet>
 *       )}
 *       <Main {...mainProps}>{children}</Main>
 *     </>
 *   );
 * }
 * ```
 *
 * If you have a large navigation panel, you can conditionally render the
 * `LayoutNav` with the `persistent` boolean returned by the hook which will
 * ensure that the DOM has rehydrated before unmounting to prevent SSR errors.
 *
 * @example Conditionally Rendering
 * ```diff
 *   const {
 *     temporary,
 * +   persistent,
 *     appBarProps,
 *     expandableNavProps,
 *     mainProps,
 *     navToggleProps,
 *     temporaryNavProps,
 *     windowSplitterProps,
 *   } = useResizableLayout({ pathname });
 *
 *   return (
 *     <>
 *       <LayoutAppBar {...appBarProps}>
 *         <Button {...navToggleProps} />
 *         <AppBarTitle>Hello, world!</AppBarTitle>
 *       </LayoutAppBar>
 * -     <LayoutNav {...expandableNavProps}>
 * -       <CustomNavigation />
 * -     </LayoutNav>
 * -     <LayoutWindowSplitter {...windowSplitterProps} />
 * +     {persistent && (
 * +       <>
 * +         <LayoutNav {...expandableNavProps}>
 * +           <CustomNavigation />
 * +         </LayoutNav>
 * +         <LayoutWindowSplitter {...windowSplitterProps} />
 * +       </>
 * +     )}
 *       {temporary && (
 *         <Sheet {...temporaryNavProps}>
 *           <CustomNavigation />
 *         </Sheet>
 *       )}
 *       <Main {...mainProps}>{children}</Main>
 *     </>
 *   )
 * ```
 *
 * @see {@link https://next.react-md.dev/getting-started/layout | Layout Demos}
 * @since 6.0.0
 */
export function useResizableLayout(
  options: ResizableLayoutOptions
): ResizableLayoutImplementation {
  const {
    navId: propNavId,
    pathname,
    appBarPosition = "fixed",
    fullHeightNav = false,
    temporaryUntil,
    defaultVisible,
    defaultExpanded,
    ...splitterOptions
  } = options;
  const navId = useEnsuredId(propNavId, "layout-nav");
  const { splitterProps, ...windowSplitter } =
    useLayoutWindowSplitter(splitterOptions);
  const expandableLayout = useExpandableLayout({
    pathname,
    appBarPosition,
    temporaryUntil,
    fullHeightNav,
    defaultExpanded,
    defaultVisible,
  });
  const { expandableNavProps, expanded } = expandableLayout;

  return {
    ...expandableLayout,
    ...windowSplitter,
    expandableNavProps: {
      id: navId,
      ...expandableNavProps,
    },
    windowSplitterProps: {
      ...splitterProps,
      "aria-controls": navId,
      appBarOffset: !fullHeightNav,
      className: (!expanded && DISPLAY_NONE_CLASS) || undefined,
    },
  };
}
