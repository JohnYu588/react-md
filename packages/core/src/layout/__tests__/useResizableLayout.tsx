import { describe, expect, it, jest } from "@jest/globals";
import type { ReactNode } from "react";
import { useAppSize } from "../../AppSizeProvider.js";
import { AppBar } from "../../app-bar/AppBar.js";
import { AppBarTitle } from "../../app-bar/AppBarTitle.js";
import { Button } from "../../button/Button.js";
import { Form } from "../../form/Form.js";
import { Radio } from "../../form/Radio.js";
import { useRadioGroup } from "../../form/useRadioGroup.js";
import { Sheet } from "../../sheet/Sheet.js";
import { drag } from "../../test-utils/drag.js";
import {
  rmdRender,
  screen,
  userEvent,
  waitFor,
} from "../../test-utils/index.js";
import { removeItemFromStorage } from "../../useLocalStorage.js";
import { isElementVisible } from "../../utils/isElementVisible.js";
import { LayoutNav } from "../LayoutNav.js";
import { LayoutWindowSplitter } from "../LayoutWindowSplitter.js";
import { Main } from "../Main.js";
import { DEFAULT_HORIZONTAL_LAYOUT_TRANSITION_CLASSNAMES } from "../useHorizontalLayoutTransition.js";
import type { ResizableLayoutOptions } from "../useResizableLayout.js";
import { useResizableLayout } from "../useResizableLayout.js";

const OFFSCREEN_CLASS = "rmd-sheet--offscreen";
const { enter: ENTER_H_CLASS, enterDone: ENTER_H_DONE_CLASS } =
  DEFAULT_HORIZONTAL_LAYOUT_TRANSITION_CLASSNAMES;

const getSizeVar = () =>
  window
    .getComputedStyle(document.documentElement)
    .getPropertyValue("--rmd-layout-size");

interface LayoutProps extends Omit<ResizableLayoutOptions, "pathname"> {
  children?: ReactNode;
}

function Layout(props: LayoutProps) {
  const { children = "Hello, world!", ...options } = props;

  const { getRadioProps, value: pathname } = useRadioGroup<string>({
    defaultValue: "/",
    name: "pathname",
  });
  const {
    appBarProps,
    mainProps,
    navToggleProps,
    temporaryNavProps,
    expandableNavProps,
    windowSplitterProps,
  } = useResizableLayout({ pathname, ...options });
  const { isPhone, isTablet, isLandscape, isDesktop } = useAppSize();
  let layout: "phone" | "tablet" | "landscapeTablet" | "desktop" = "phone";
  if (isTablet) {
    layout = isLandscape ? "landscapeTablet" : "tablet";
  } else if (isDesktop) {
    layout = "desktop";
  }

  return (
    <>
      <AppBar {...appBarProps}>
        <Button {...navToggleProps} />
        <AppBarTitle>Hello, world!</AppBarTitle>
      </AppBar>
      <LayoutNav {...expandableNavProps}>
        {/* pretend like these are navigation items */}
        <Form>
          <Radio {...getRadioProps("/")} label="Home" />
          <Radio {...getRadioProps("/1")} label="Route 1" />
          <Radio {...getRadioProps("/2")} label="Route 2" />
        </Form>
      </LayoutNav>
      <LayoutWindowSplitter {...windowSplitterProps} />
      {isPhone && (
        <Sheet {...temporaryNavProps}>
          {/* pretend like these are navigation items */}
          <Form>
            <Radio {...getRadioProps("/")} label="Home" />
            <Radio {...getRadioProps("/1")} label="Route 1" />
            <Radio {...getRadioProps("/2")} label="Route 2" />
          </Form>
        </Sheet>
      )}
      <Main {...mainProps}>
        {children}
        <div data-testid="layout">{layout}</div>
      </Main>
    </>
  );
}

describe("useResizableLayout", () => {
  afterEach(() => {
    removeItemFromStorage({ key: "navWidth" });
  });

  it("should allow you to create a resizable layout quickly", async () => {
    const user = userEvent.setup();
    rmdRender(<Layout />);

    const layout = screen.getByTestId("layout");
    const appBar = screen.getByRole("banner");
    const main = screen.getByRole("main");
    const navToggle = screen.getByRole("button", { name: "Navigation" });
    const nav = screen.getByRole("navigation", { name: "Navigation" });
    const windowSplitter = screen.getByRole("separator", {
      name: "Resize Navigation",
    });

    expect(layout).toHaveTextContent("desktop");
    expect(isElementVisible(nav)).toBe(false);
    expect(nav).toHaveClass(OFFSCREEN_CLASS);
    expect(appBar).toMatchSnapshot();
    expect(main).toMatchSnapshot();
    expect(isElementVisible(windowSplitter)).toBe(false);
    expect(windowSplitter).toMatchSnapshot();
    expect(getSizeVar()).toBe("256px");
    expect(() => screen.getByRole("dialog")).toThrow();

    await user.click(navToggle);
    expect(isElementVisible(nav)).toBe(true);
    expect(main).toHaveClass(ENTER_H_CLASS);
    await waitFor(() => {
      expect(nav).not.toHaveClass(OFFSCREEN_CLASS);
    });
    expect(main).toHaveClass(ENTER_H_DONE_CLASS);

    expect(isElementVisible(windowSplitter)).toBe(true);
    expect(windowSplitter).toMatchSnapshot();

    jest.spyOn(windowSplitter, "getBoundingClientRect").mockReturnValue({
      ...document.body.getBoundingClientRect(),
      left: 256,
      width: 8,
      top: 0,
      height: window.innerHeight,
    });

    await drag(windowSplitter, {
      to: { x: 286, y: 32 },
    });

    expect(getSizeVar()).toBe("286px");
    expect(windowSplitter).toMatchSnapshot();

    await drag(windowSplitter, {
      to: { x: 0, y: 50 },
    });
    expect(getSizeVar()).toBe("96px");
    expect(windowSplitter).toMatchSnapshot();
  });
});
