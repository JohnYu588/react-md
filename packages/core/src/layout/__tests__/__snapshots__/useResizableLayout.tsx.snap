// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useResizableLayout should allow you to create a resizable layout quickly 1`] = `
<header
  class="rmd-app-bar rmd-app-bar--primary rmd-app-bar--fixed rmd-app-bar--top rmd-app-bar--elevated rmd-app-bar--scrollbar-offset"
>
  <button
    aria-label="Navigation"
    class="rmd-button rmd-button--icon"
    type="button"
  >
    <span
      aria-hidden="true"
      class="rmd-icon rmd-icon--font material-icons"
    >
      menu
    </span>
  </button>
  <h6
    class="rmd-typography rmd-typography--headline-6 rmd-app-bar-title rmd-app-bar-title--no-wrap"
  >
    Hello, world!
  </h6>
</header>
`;

exports[`useResizableLayout should allow you to create a resizable layout quickly 2`] = `
<main
  class="rmd-layout-main rmd-layout-main--offset-v"
  id="main-:r5:"
>
  Hello, world!
  <div
    data-testid="layout"
  >
    desktop
  </div>
</main>
`;

exports[`useResizableLayout should allow you to create a resizable layout quickly 3`] = `
<button
  aria-controls="layout-nav-:r0:"
  aria-label="Resize Navigation"
  aria-valuemax="100"
  aria-valuemin="0"
  aria-valuenow="32"
  class="rmd-window-splitter rmd-window-splitter--h rmd-layout-splitter rmd-layout-splitter--offset-v rmd-layout-splitter--responsive rmd-display-none"
  id="splitter-:r1:"
  role="separator"
  type="button"
/>
`;

exports[`useResizableLayout should allow you to create a resizable layout quickly 4`] = `
<button
  aria-controls="layout-nav-:r0:"
  aria-label="Resize Navigation"
  aria-valuemax="100"
  aria-valuemin="0"
  aria-valuenow="32"
  class="rmd-window-splitter rmd-window-splitter--h rmd-layout-splitter rmd-layout-splitter--offset-v rmd-layout-splitter--responsive"
  id="splitter-:r1:"
  role="separator"
  type="button"
/>
`;

exports[`useResizableLayout should allow you to create a resizable layout quickly 5`] = `
<button
  aria-controls="layout-nav-:r0:"
  aria-label="Resize Navigation"
  aria-valuemax="100"
  aria-valuemin="0"
  aria-valuenow="38"
  class="rmd-window-splitter rmd-window-splitter--h rmd-layout-splitter rmd-layout-splitter--offset-v rmd-layout-splitter--responsive"
  id="splitter-:r1:"
  role="separator"
  type="button"
/>
`;

exports[`useResizableLayout should allow you to create a resizable layout quickly 6`] = `
<button
  aria-controls="layout-nav-:r0:"
  aria-label="Resize Navigation"
  aria-valuemax="100"
  aria-valuemin="0"
  aria-valuenow="0"
  class="rmd-window-splitter rmd-window-splitter--h rmd-layout-splitter rmd-layout-splitter--offset-v rmd-layout-splitter--responsive"
  id="splitter-:r1:"
  role="separator"
  type="button"
/>
`;
